### ISSUE TITLE
Use Mozilla Readability instead of body.innerText

### ISSUE THREAD

**Issue by bwbroersma** [WebMemex/webmemex-extension#2]
Instead of [`body.innerText`](https://github.com/WebMemex/memextension/blob/8fdff6c167b618793d46ef8836f6ec615c499b72/src/activitylogger/content_script/index.js#L7) use [Mozilla Readability](https://github.com/mozilla/readability).

**Comment by blackforestboi** [comment]
@bwbroersma 

AFAIK body.innerText is a more inclusive way of capturing text content, as it is done by the browser itself, after rendering.
Readability is parsing the html and has a lot of blind spots for text content, that is not part of the main text in an article. This is because it is optimised for articles, not for web pages in general.

**Comment by bwbroersma** [comment]
@oliversauter: I know Readability is limiting the content, but that is partly what I want. It would be better to index both (filtered context and full context) so it can be used during the ranking of the search. If Readability doesn't find a good candidate [it will use the body as last resort](https://github.com/mozilla/readability/blob/cec7d90a14b977ea80df12864c311b160f74dc34/Readability.js#L829).

**Comment by blackforestboi** [comment]
So i understand it correctly, that readability would throw out the same content as body.innerText, if it is not an article (where no clear main text is available)?

**Comment by bwbroersma** [comment]
You can try this in your dev console:
```js
const scriptNode = document.createElement('script');
scriptNode.src = 'https://rawgit.com/mozilla/readability/master/Readability.js';
document.body.appendChild(scriptNode);
const loc = document.location;
const uri = {
  spec: loc.href,
  host: loc.host,
  prePath: loc.protocol + "//" + loc.host,
  scheme: loc.protocol.substr(0, loc.protocol.indexOf(":")),
  pathBase: loc.protocol + "//" + loc.host + loc.pathname.substr(0, loc.pathname.lastIndexOf("/") + 1)
};
scriptNode.onload = () => {
    const article = new Readability(uri, document.cloneNode(true)).parse();
    console.log(article.content);
};
```
It only works on sites without strict Content Security Policy (so it doesn't work on github), on [hackernews](https://news.ycombinator.com/) it generates some interesting HTML (`tr`/`td` without `table`).

**Comment by blackforestboi** [comment]
I see 2 problems: 
1. It gives me a web compilable html code, but not the plain text that is necessary for the search.
Additional, error risky, steps would need to be taken to extract the text

2.  body.innertext captures everything on the page, which is good for now, as we rather have too many words as too less. Especially with not everything being an article that can be parsed with readability. 
Try your approach to this page and compare it to body.innerText as an example:
http://appliedmythology.blogspot.de/

Sure, you could end up with too many (or not precise enough) search results, but that can be solved by filters that need to be implemented anyway.  (for example filtering by the domain, some time frame or other stuff)

**Comment by bwbroersma** [comment]
Improvement test case on a clean extension install:
1. goto a Wikipedia page with translations like [World Wide Web](https://en.wikipedia.org/wiki/World_Wide_Web)
2. search WebMemex for `Svenska`

With the [Readability fix](https://github.com/WebMemex/memextension/pull/3/commits/fdd019c62028b9ae66c9f107cabdba7b4fe77f08) this will result in no results, instead of the Wikipedia page, since the languages are outside of the content.

About your issues:
1. Just creating a DOM Element and setting its `innerHTML` to the .content of the Readability parse output works for me.
2. Different ranking should be used for content & page text, your Blogspot sample is very interesting, I don't get a Reader view in Firefox, but the `content` is indeed way to limited, I will check how Firefox fixed the Blogspot issue.

**Comment by Treora** [comment]
I would very much like to do some more sophisticated web page interpretation, including extraction of the main text(s). Mozilla's Readability seems a suitable tool for this, though I am not so convinced of its current code quality. I just sent a [PR](https://github.com/mozilla/readability/pull/340) to make it CommonJS-compatible, and [tried it out in the extension](https://github.com/WebMemex/memextension/tree/readability).

**Comment by bwbroersma** [comment]
@Treora: since Firefox is using it, why isn't Firefox displaying the 'Enter Reader View' icon on sites like Blogspot, but is Readability not failing on the same site? (idem for [HackerNews](https://news.ycombinator.com), where Firefox has no Reader icon) Adding the same logic would 'fix' https://github.com/WebMemex/memextension/issues/2#issuecomment-272973980, since I actually never encountered a fully broken Reader View in Firefox..

**Comment by bwbroersma** [comment]
Ok found it, see [`isProbablyReaderable(doc)`](https://github.com/mozilla/gecko-dev/blob/a712d69adb9b2588f88aff678216b2be94d3719c/toolkit/components/reader/ReaderMode.jsm#L169-L191) in the Firefox code, next to [a hard coded blacklist](https://github.com/mozilla/gecko-dev/blob/a712d69adb9b2588f88aff678216b2be94d3719c/toolkit/components/reader/ReaderMode.jsm#L379-L412), it uses the [`isProbablyReaderable`](https://github.com/mozilla/readability/blob/cec7d90a14b977ea80df12864c311b160f74dc34/Readability.js#L1767-L1824) method of Readability, which would filter out the Blogspot case. They also do some caching, might be sensible to copy that too?

**Comment by Treora** [comment]
Oh I forgot to close this. We included Readability in 6b0b2c8. Of course `body.innerText` is used as a fallback. I just [changed it](https://github.com/WebMemex/memextension/blob/6dc134d3d583af4c9501bfeedde8a99d4b6529ec/src/page-analysis/content_script/extract-page-text.js#L34) so that it does remember *which* type of text it has saved. Also it now keeps the whole object that Readability returns, which may be more than needed but things can be forgotten again later.

**Comment by Treora** [comment]
Readability was removed again in a0b8255 (#101), as we don't use it anymore for viewing stored pages, it is too unreliable to not also store the full text, and does not seem worth the added complexity.
